### PR TITLE
[IM] Notify report engine when adding or enabling endpoints

### LINKED ISSUE #10795: We should notify about the fact that all the attributes for
https://github.com/project-chip/connectedhomeip/blob/403c7cc94fd5d1ac71a4c1c1490c77c4bb1308ea/src/app/util/attribute-storage.cpp#L1002-L1012

---

###### This issue was generated by [todo](https://todo.jasonet.co) based on a `TODO` comment in 403c7cc94fd5d1ac71a4c1c1490c77c4bb1308ea when #10793 was merged. cc @bzbarsky-apple.

### LINKED ISSUE #15120: Reject Write Requests if one is already outstanding
On the server, we should reject Write requests if one is already outstanding. This can only happen if the outstanding write is chunked and more messages to convey the remaining data are forthcoming on the exchange.

Ideally, we would achieve serialization at the attribute level, but that isn't easy to achieve.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
Fixes #10795 

#### Change overview

Notify report engine for newly enabled endpoint

#### Testing

Added unit test for dynamic endpoint and disabling / re-enabling endpoint

### CODE REVIEW

File: src/app/util/ember-compatibility-functions.cpp [hunk 1]
```diff
@@ -1039,3 +1039,22 @@ void MatterReportingAttributeChangeCallback(const ConcreteAttributePath & aPath)
...
+    assertChipStackLockedByCurrentThread();
+
+    ClusterInfo info;
+    info.mEndpointId = endpoint;
+
+    // We are adding or enabling a whole endpoint, in this case, we do not touch the cluster data version.
+
+    InteractionModelEngine::GetInstance()->GetReportingEngine().SetDirty(info);
+
+    // Schedule work to run asynchronously on the CHIP thread. The scheduled work won't execute until the current execution context
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:1056]: Please update to tip.  SetDirty() handles all that itself on tip.

File: src/controller/tests/TestReadChunking.cpp [hunk 2]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+                                   app::ReadClient::InteractionType::Subscribe);
+
+        NL_TEST_ASSERT(apSuite, readClient.SendRequest(readParams) == CHIP_NO_ERROR);
+
+        for (int j = 0; j < 3 && !readCallback.mOnSubscriptionEstablished; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:484]: This should just have a single `ctx.DrainAndServiceIO();` call. There should be no need for manual `Run()` calls or a loop here.
```suggestion
        ctx.DrainAndServiceIO();
```

File: src/controller/tests/TestReadChunking.cpp [hunk 3]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+
+        // Enable the new endpoint
+        emberAfSetDynamicEndpoint(0, kTestEndpointId4, &testEndpoint4, 0, 0, Span<DataVersion>(dataVersionStorage));
+
+        for (int j = 0; j < 2 && !readCallback.mOnReportEnd; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:498]: Again, there should be no manual `Run()` call and just a single `ctx.DrainAndServiceIO();` here.
```suggestion
        ctx.DrainAndServiceIO();
```

File: src/controller/tests/TestReadChunking.cpp [hunk 4]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+
+        // Ensure we have received the report, we do not care about the initial report here.
+        // ClientGeneratedCommandList / ServerGeneratedCommandList / AttributeList attribute are not include in
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:501]:
```suggestion
        // ClientGeneratedCommandList / ServerGeneratedCommandList / AttributeList attribute are not included in
```

File: src/controller/tests/TestReadChunking.cpp [hunk 5]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+
+        // Disable the new endpoint
+        emberAfEndpointEnableDisable(kTestEndpointId4, false);
+
+        for (int j = 0; j < 2 && !readCallback.mOnReportEnd; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:519]:
```suggestion
        ctx.DrainAndServiceIO();
```

File: src/controller/tests/TestReadChunking.cpp [hunk 6]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+
+        // We may receive some attribute reports for descriptor cluster, but we does not care about it for now.
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:521]:
```suggestion
        // We may receive some attribute reports for descriptor cluster, but we do not care about it for now.
```

File: src/controller/tests/TestReadChunking.cpp [hunk 7]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+        readCallback.mAttributeCount = 0;
+        readCallback.mOnReportEnd    = false;
+
+        emberAfEndpointEnableDisable(kTestEndpointId4, true);
+        for (int j = 0; j < 2 && !readCallback.mOnReportEnd; j++)
+        {
+            ctx.DrainAndServiceIO();
+            chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+            ctx.DrainAndServiceIO();
+        }
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:534]:
```suggestion
        ctx.DrainAndServiceIO();
```

File: src/controller/tests/TestReadChunking.cpp [hunk 8]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+        // testClusterAttrsOnEndpoint4.
+        NL_TEST_ASSERT(apSuite, readCallback.mAttributeCount == ArraySize(testClusterAttrsOnEndpoint4) + 3);
+
+        // We have received all report data.
+        NL_TEST_ASSERT(apSuite, readCallback.mOnReportEnd);
+    }
+
+    chip::test_utils::SleepMillis(secondsToMilliseconds(2));
+
+    // Destroy the read client will terminate the subscription transaction.
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:547]:
```suggestion
    // Destroyin the read client will terminate the subscription transaction.
```

File: src/controller/tests/TestReadChunking.cpp [hunk 9]
```diff
@@ -427,12 +445,125 @@ void TestCommandInteraction::TestBadChunking(nlTestSuite * apSuite, void * apCon
...
+    for (int j = 0; j < 2 && !readCallback.mOnReportEnd; j++)
+    {
+        ctx.DrainAndServiceIO();
+        chip::app::InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
+        ctx.DrainAndServiceIO();
+    }
```
Comment (bzbarsky-apple) [inline src/controller/tests/TestReadChunking.cpp:553]:
```suggestion
    ctx.DrainAndServiceIO();
```

File: src/controller/tests/TestReadChunking.cpp (hunk 8 above)
Comment (erjiaqing) [reply src/controller/tests/TestReadChunking.cpp:524]:
```suggestion
    // Destroying the read client will terminate the subscription transaction.
```

File: src/app/util/ember-compatibility-functions.cpp (hunk 1 above)
Comment (erjiaqing) [reply src/app/util/ember-compatibility-functions.cpp]: Thanks for notice.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**